astropy/utils/tests/test_data.py

```diff
@@ -372,3 +376,37 @@ def test_is_url_in_cache():
 
     download_file(TESTURL, cache=True, show_progress=False)
     assert is_url_in_cache(TESTURL)
+
+
+@contextlib.contextmanager
+def TemporaryTempDir():
+    """Temporarily redirect all temporary files to a temporary directory"""
+    tempdir = tempfile.mkdtemp()
+    old_tempdir = tempfile.tempdir
+    tempfile.tempdir = tempdir
+    try:
+        yield tempdir
+    finally:
+        tempfile.tempdir = old_tempdir
+        shutil.rmtree(tempdir)
+
+
+def test_get_readable_fileobj_cleans_up_temporary_files():
```

Comment:
Thanks for writing this test--you actually made yourself do too much work :(  You can use a couple facilities built into py.test as fixtures that would make this easier:

```python
def test_get_readable_fileobj_cleans_up_temporary_files(tmpdir, monkeypatch):
    monkeypatch.setattr(tempfile, 'tempdir', str(tmpdir)

    # Do the test more or less like you have before

    assert len(tmpdir.listdir()) == 0
```